### DISCUSSION TITLE
Best Practices with Virtualenv? (venv/pipenv/poetry etc.) [Possible Mise Improvement]

### DISCUSSION THREAD

**Discussion by BuildBackBuehler** [jdx/mise#1835] [Q&A]
Hey all, 

Been using Mise for a surprisingly long time considering the questions I'd like to ask. I suppose I widdled it down to the least dumber-est for yas

I've relied on pipenv for a long time because I hate Conda with a passion and I find Pip without the smart package versioning to be too big of an annoyance. Although, I guess they took care of that with Pip 24(?)...maybe earlier and I just didn't notice/know haha. 

I was coming around with Conda using Mamba, but its still a bit obtuse for my tastes. 

The thing is, I just want the one that'll allow me to be the laziest I can be aka trust there won't be errors, the most intuitive & automatic. With new Mise features, I think I can probably achieve that with any of 'em honestly. I figure worst case scenario I'd write a wee script to take whichever the extra inch.

So it seems at the moment with Mise .venv is the only option for automatic creation of .venv upon folder entry. I'm not sure if I've jerry-rigged it to work with a Pipenv created virtualenv or if it is Pipenv working within .venv

`
[env]
PIPENV_VENV_IN_PROJECT=1
PIPENV_FILE = "~/local/gitrepos/comfyui/pipfile"
PIPENV_MAX_DEPTH = "7"
_.python.venv = { path = "~/gitrepos/comfyui/.venv", create = true }`

Now, my 1 problem with using Pipenv is the fact that I can't setup a default_python_package file [I don't think?] and same would go for virtenv_def_pack. Though I suppose that could be a nifty improvement (if in the file a user doesn't write their package provider, it defaults to pip, but otherwise you could write in "package_provider=pipenv/poetry")

The End!

...or is it? Random Qs, rather than make another discussion...

Is it best practice to disregard adding the Mise tools into my .zshenv file (where I've got a bunch of if [[ directory + python command ]]; then path = (`path to python/bin` \\ `path`))? I imagine based on the docs, its taken care of.

Also wondering how I can link Mise's Python, or whatever, into Brew. Do I just do the ol' Brew install [Mise Python Path] --cask trick or would that mess everything up? I ask this because I have had difficulties with compiling packages with Clang, even though it is installed via /usr/bin/clang and brew/llvm/clang. It may have been the packages being uncompile-able, though. 

Thanks for any help!

![part0](https://github.com/jdx/mise/assets/116780517/c10cd4ed-ce44-4b1e-b9b4-77291696a3af)
A pup pic to thank you for your time

**Comment by Marukome0743** [comment]
For current mise versions, the simplest approach depends on whether the project uses uv.

For a uv project, I would recommend letting uv own dependency resolution and the virtual environment:

```toml
[tools]
python = "3.13"
uv = "latest"

[settings]
python.uv_venv_auto = "create|source"
```

After `uv sync` or `uv lock` has created a `uv.lock`, mise detects the project and creates or activates uv’s `.venv`. It also respects `UV_PROJECT_ENVIRONMENT` when a different environment path is configured.

For Pipenv, Poetry, or a plain `venv`, let that tool create the environment and configure mise to activate the resulting path:

```toml
[tools]
python = "3.13"

[env]
_.python.venv = ".venv"
```

If mise itself should create the environment:

```toml
[env]
_.python.venv = { path = ".venv", create = true }
```

When Pipenv owns the environment, `PIPENV_VENV_IN_PROJECT=1` is useful, but I would avoid combining it with `create = true`: otherwise mise may create a plain venv before Pipenv initializes its own environment.

Additional `PATH` handling in `.zshenv` should not be necessary when using `mise activate` or `mise exec`. I would also not try to link a mise-managed Python installation into Homebrew. Homebrew can provide native build dependencies such as libraries or LLVM, while mise should continue to manage the Python interpreter separately.

The older `[tools].python.virtualenv` option is now deprecated; `_.python.venv` and `python.uv_venv_auto` are the current mechanisms.

https://mise.en.dev/lang/python.html

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
